torch/distributed/fsdp/__init__.py

```diff
@@ -17,3 +17,23 @@
     StateDictSettings,
     StateDictType,
 )
+
+__all__ = [
+    "BackwardPrefetch",
+    "CPUOffload",
+    "FlatParameter",
```

Comment:
Can we avoid exporting `FlatParameter`? This is an internal implementation that users should not have to directly use.

Reply:
When I previously didn't include it: https://github.com/pytorch/pytorch/actions/runs/6279593214/job/17055789987#step:15:2112

>FAILED [2.8867s] test_public_bindings.py::TestPublicBindings::test_correct_module_names - AssertionError: False is not true : All the APIs below do not meet our guidelines for public API from https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation.
Make sure that everything that is public is expected (in particular that the module has a properly populated `__all__` attribute) and that everything that is supposed to be public does look public (it does not start with `_` and has a `__module__` that is properly populated).
> 
> torch.distributed.fsdp.FlatParameter:
>  - Is NOT public: it is not inside the module's (`torch.distributed.fsdp`) `__all__`
>  - Does look public: it does look public because it follows the rules from the doc above (does not start with `_` and has a proper `__module__`).
>  - You can do either of these two things to fix this problem:
>    - To make it public: add it from the modules's (`torch.distributed.fsdp`) `__all__`
>    - To make it NOT look public: make its name start with `_`

What is the alternative that we would do instead?

Reply:
@albanD Any thoughts about this?

Reply:
I wonder if it is already too late for us since `flat_param.py` is public and `named_parameters()` returns `FlatParameter` for `use_orig_params=False` 😞

Reply:
I guess we should make `flat_param.py` be private and make it a BC breaking change? Given that FSDP is still beta I feel it's worth making this change. `named_parameters()` return FlatParam when `use_orig_params=False` should be fine, as long as user don't use `FlatParam` directly, this is risky if we expose as public, and when we want to switch the underlying implementation or delete it in future, it's gonna be more problematic

Reply:
@b-chu do you want to take on this work in #108914?

Reply:
I've added the 2 additional export aliases from #108914 and applied the change @wanchaol suggested here